tv_apps/tv-deck/js/tv_deck.js

```diff
@@ -0,0 +1,538 @@
+    }
+
+    this.playingTuner.tuner.setCurrentSource(this.playingSourceType)
+      .then(function() {
+        this.setPlayingChannel(callback);
+      }.bind(this), function() {
+        alert('Source not found');
+      });
+  };
+
+  proto.setPlayingChannel = function td_setPlayingChannel(callback) {
+    this.playingSource.source.setCurrentChannel(this.playingChannelNumber)
+      .then(function() {
+        /* jshint maxlen:false */
+        var newStream = this.playingTuner.tuner.stream;
+        if (this.tvStreamElement.src !== newStream) {
+          this.tvStreamElement.src = newStream;
+          this.tvStreamElement.play();
+        }
+        if (callback) {
+          callback();
+        }
+      }.bind(this), function() {
+        alert('Channel not found');
+      });
+  };
+
+  proto._compareChannel = function td__compareChannel(channelA, channelB) {
+    var dashIndexA = channelA.indexOf('-');
+    var dashIndexB = channelB.indexOf('-');
+    var numberA = (dashIndexA >= 0) ?
+                        channelA.substring(0, dashIndexA) : channelA;
+    var numberB = (dashIndexB >= 0) ?
+                        channelB.substring(0, dashIndexB) : channelB;
+    numberA = parseInt(numberA, 10);
+    numberB = parseInt(numberB, 10);
+
+    if (numberA === numberB) {
+      if (dashIndexA === -1 && dashIndexB === -1) {
+        return 0;
+      } else if (dashIndexA !== -1 && dashIndexB === -1) {
+        return 1;
+      } else if (dashIndexA === -1 && dashIndexB !== -1) {
+        return -1;
+      }
+
+      channelA = channelA.substring(dashIndexA + 1);
+      channelB = channelB.substring(dashIndexB + 1);
+      return this._compareChannel(channelA, channelB);
+    }
+    return numberA - numberB;
+  };
+
+  proto.updateChannelInfo = function td_updateChannelInfo(title, number) {
+    this.channelTitle.textContent = title || this.playingChannel.channel.name;
+    this.channelNumber.textContent = number || this.playingChannelNumber;
+  };
+
+  proto.setHash = function td_setHash() {
+    if (this.currentHash === window.location.hash.substring(1)) {
+      // When initizlizing TVDeck, if either URL hash or localStorage is not
+      // null, the URL will be set to the same value in _onScanningCompleted
+      // function. However, we still have to call onHashChange function in order
+      // to switch to the correct channel.
+      this._onHashChange();
+    }
+    window.location.hash = this.currentHash;
+  };
+
+  /**
+   * Up and down keys are for switching channels, where left and right keys are
+   * for navigating buttons in button-group.
+   */
+  proto._onSwitch = function td__onSwitch(direction) {
+    var channelList = this.playingSource.channels;
+    var channelIndex = this.playingSource
+                           .channelIndexHash[this.playingChannelNumber];
+
+    switch(direction) {
+      case 'up':
+        channelIndex += 2;
+       /* falls through */
+      case 'down':
+        channelIndex --;
```

Comment:
press up `channelIndex += 2` and `channelIndex --`?? The code here is so confusing. I can get your point. But please prevent to use such kind of code.